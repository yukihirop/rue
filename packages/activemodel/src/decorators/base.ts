import i18n from '@/locales';

// types
import type * as t from './types';
import type * as lt from '@/locales';

export const RueCheck = (opts?: t.RueCheckOptions) => (target) => {
  if (!opts) opts = { uniqueKey: true, i18nConfig: true };
  try {
    if (opts.uniqueKey) {
      target.checkUniqueKey();
    }
    if (opts.i18nConfig) {
      target.checkI18nConfig();
    }
  } catch (err) {
    throw err;
  }
  return target;
};

/**
 * @see https://www.i18next.com/overview/api#addresourcebundle
 */
const RueSetupI18n = (target) => {
  const { options, resources } = target.i18nConfig() as lt.I18nConfig;
  const { lng } = options;
  const namespace = 'translation';
  i18n.changeLanguage(lng);
  i18n.addResourceBundle(lng, namespace, resources[lng], true, true);
  return target;
};

export const RueSetup = (target) => {
  return RueSetupI18n(target);
};
