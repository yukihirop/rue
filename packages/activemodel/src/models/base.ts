// rue packages
import { ActiveSupport$Base as Support } from '@rue/activesupport';

// third party
import flatten from 'obj-flatten';

// local
import { ActiveModel$Impl } from './impl';

// types
import type * as t from './types';
import type { Validations$Errors } from '@/models/modules/validations';

export class ActiveModel$Base extends ActiveModel$Impl {
  public errors: Validations$Errors;

  static getProperties(): string[] {
    return Support.getProperties();
  }

  constructor(data: t.Params = {}) {
    super();
    this.errors = {};
    if (!data) data = {};
    Object.keys(data).forEach((key) => {
      (this as any)[key] = data[key];
    });
  }

  toObj(opts?: { flat: boolean }): t.Params {
    opts = opts == undefined ? { flat: false } : opts;
    const instance = Object.create(this);
    const obj = Object.assign({}, instance.__proto__);
    return opts.flat ? Object.assign(flatten(obj), { errors: obj.errors }) : obj;
  }

  inspect(): string {
    return Support.inspect(this);
  }
}
