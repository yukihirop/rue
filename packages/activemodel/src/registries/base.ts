// rue packages
import { Registry } from '@rue/activesupport';

// types
import type * as st from '@rue/activesupport';
import type * as t from './types';

// define singletons
const registryForValidations = new Registry<t.Validations>('Validations');

const innerRegistryForUniqueKeys = new Registry<string[]>('RueClassNames');
class registryForUninqueKeys {
  private static readonly klassName = 'RueUniqueKey';
  private static readonly key = 'all';

  static create(val: st.Registry$Value) {
    innerRegistryForUniqueKeys.create(this.klassName, this.key, val);
  }

  static update(val: st.Registry$Value) {
    innerRegistryForUniqueKeys.update(this.klassName, this.key, val);
  }

  static read<T>(ensureType?: st.Registry$Type): T {
    return innerRegistryForUniqueKeys.read<T>(this.klassName, this.key, ensureType);
  }
}

// prettier-ignore
export { registryForValidations, registryForUninqueKeys };
