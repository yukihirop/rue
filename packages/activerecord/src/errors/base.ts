// rue packages
import { ActiveModel$Error } from '@ruejs/activemodel';

// local
import { ErrMessages } from './code';

// types
import * as t from './types';

const NAMESPACE = '@ruejs/activerecord';

export class ActiveRecord$Error extends ActiveModel$Error {}

export function createErrObj({
  code,
  params,
  message,
}: {
  code: string;
  params?: any;
  message?: string;
}): t.ErrObj {
  let msg = ErrMessages[code].message;
  let calcMessage: string = msg instanceof Function ? msg(params) : msg!;
  const maybeOverrideMessage = message != undefined ? message : calcMessage;

  const err = new ActiveRecord$Error(maybeOverrideMessage);
  err.namespace = NAMESPACE;
  err.code = code;
  err.message = maybeOverrideMessage!;

  return err;
}
