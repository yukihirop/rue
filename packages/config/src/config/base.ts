// third party
import { js as beautify } from 'js-beautify';
import stringifyObject from 'stringify-object';

// builtin
import path from 'path';
import fs from 'fs';

// types
import * as t from './types';

const projectRoot = require('pkg-dir').sync() || process.cwd();

export class Config$Base {
  static fileName = 'rue.config.js';
  static default: t.RueConfig = {
    backend: {
      mock_server: {
        loadData: ['./backend/mock_server/data/**/*.{js,json}'],
        dist: {
          db: './backend/mock_server/db.json',
          routes: './backend/mock_server/routes.json',
        },
      },
    },
    cli: {
      commands: {
        console: {
          nodeREPL: {
            prompt: '🍛 > ',
            useColors: true,
          },
        },
      },
    },
    repl: {
      ruePackageRootPath: undefined,
      actions: {
        ancestors: 'ancs',
        descriptors: 'desc',
        loadedModules: 'loaded',
        propertyList: 'lp',
        methodList: 'ls',
        prototype: 'proto',
        definition: 'show',
      },
      dotEnvPath: './.env',
      loadModules: [
        'src/**/{forms,models,records}/**/*.{js,ts}',
        '!src/**/__tests__/*.test.{js,ts}',
      ],
      moduleAliases: {
        '@': './src',
      },
    },
  };
  static defaultJS: string = beautify(Config$Base.getJSConfig(Config$Base.default), {
    indent_size: 2,
    space_in_empty_paren: true,
    unescape_strings: false,
  });

  static all(): t.RueConfig {
    const loadedCofigPath = path.join(projectRoot, this.fileName);
    let config;

    if (fs.existsSync(loadedCofigPath)) {
      config = require(loadedCofigPath);
    } else {
      config = this.default;
    }

    return config;
  }

  static nodeREPL(): t.RueNodeREPLConfig {
    return this.all()['cli']['commands']['console']['nodeREPL'];
  }

  static rueREPL(): t.RueRueREPLConfig {
    return this.all()['repl'];
  }

  static get backend(): t.RueBackendConfig {
    return this.all()['backend'];
  }

  // https://stackoverflow.com/a/11233515/9434894
  // https://stackoverflow.com/a/53990000/9434894
  private static getJSConfig(config: t.RueConfig): string {
    return `
    module.exports = ${stringifyObject(config)};
    `;
  }
}
