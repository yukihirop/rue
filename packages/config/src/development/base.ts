// types
import * as t from './types';

export class Config$Base {
  static fileName = 'rue.config.js';
  static default: t.RueConfig = {
    backend: {
      mock_server: {
        loadData: ['./backend/mock_server/data/**/*.{js,json}'],
        dist: {
          db: './backend/mock_server/db.json',
          routes: './backend/mock_server/routes.json',
        },
      },
    },
    cli: {
      commands: {
        console: {
          nodeREPL: {
            prompt: '🍛 > ',
            useColors: true,
          },
        },
      },
    },
    repl: {
      ruePackageRootPath: undefined,
      actions: {
        ancestors: 'ancs',
        descriptors: 'desc',
        loadedModules: 'loaded',
        propertyList: 'lp',
        methodList: 'ls',
        prototype: 'proto',
        definition: 'show',
      },
      dotEnvPath: './.env',
      loadModules: [
        'src/**/{forms,models,records}/**/*.{js,ts}',
        '!src/**/__tests__/*.test.{js,ts}',
      ],
      moduleAliases: {
        '@': './src',
      },
    },
  };
  static defaultJS: string = require('js-beautify').js(
    Config$Base.getJSConfig(Config$Base.default),
    {
      indent_size: 2,
      space_in_empty_paren: true,
      unescape_strings: false,
    }
  );

  static all(): t.RueConfig {
    /**
     * @description
     * If you declare it with import, an error will occur when this class is loaded in the browser.
     * Both fs and path can only be used on the server side.
     */
    const fs = require('fs');
    const path = require('path');
    const projectRoot = require('pkg-dir').sync() || process.cwd();
    const loadedCofigPath = path.join(projectRoot, this.fileName);
    let config;

    if (fs.existsSync(loadedCofigPath)) {
      config = require(loadedCofigPath);
    } else {
      config = this.default;
    }

    return config;
  }

  static nodeREPL(): t.RueNodeREPLConfig {
    return this.all()['cli']['commands']['console']['nodeREPL'];
  }

  static rueREPL(): t.RueRueREPLConfig {
    return this.all()['repl'];
  }

  static get backend(): t.RueBackendConfig {
    return this.all()['backend'];
  }

  // https://stackoverflow.com/a/11233515/9434894
  // https://stackoverflow.com/a/53990000/9434894
  private static getJSConfig(config: t.RueConfig): string {
    const stringifyObject = require('stringify-object');
    return `
    module.exports = ${stringifyObject(config)};
    `;
  }
}
