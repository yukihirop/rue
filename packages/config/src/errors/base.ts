import { ErrMessages } from './code';

// types

import * as t from './types';

const NAMESPACE = '@ruejs/activemodel';

// The namespace is given to the name so that "ActiveModel$Error" is displayed when viewed on the console.
export class RueConfig$Error extends Error implements t.ErrObj {
  private _namespace?: string;
  private _code?: string;
  private _message?: string;

  get namespace(): string {
    return this._namespace || '';
  }

  set namespace(val: string) {
    this._namespace = val;
  }

  get code(): string {
    return this._code || '';
  }

  set code(val: string) {
    this._code = val;
  }

  get message(): string {
    return this._message || '';
  }

  set message(val: string) {
    this._message = val;
  }
}

export function createErrObj({
  code,
  params,
  message,
}: {
  code: string;
  params?: any;
  message?: string;
}): t.ErrObj {
  let { message: msg } = ErrMessages[code];
  let calcMessage: string = msg instanceof Function ? msg(params) : msg!;
  const maybeOverrideMessage = message != undefined ? message : calcMessage;

  const err = new RueConfig$Error(maybeOverrideMessage);
  err.namespace = NAMESPACE;
  err.code = code;
  err.message = maybeOverrideMessage!;

  return err;
}
