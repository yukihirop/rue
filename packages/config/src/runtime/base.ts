// functions
import { errObj, ErrCodes } from '@/errors';

// types
import type * as t from './types';

const SUPPORT_LANGUAGES = ['ja', 'en'];

class RueRuntimeConfig$Base {
  public i18n: t.RueRuntimeConfig['i18n'];

  configure(data: t.RueRuntimeConfig) {
    this.i18n = data['i18n'];
    this.check();
  }

  check() {
    if (!SUPPORT_LANGUAGES.includes(this.i18n.options.lng)) {
      throw errObj({
        code: ErrCodes.CONFIG_IS_INVALID,
        params: {
          body: `'i18n.options.lng' must be included in [${SUPPORT_LANGUAGES.join(', ')}]`,
        },
      });
    }

    if (
      this.i18n.options.fallbackLng &&
      !SUPPORT_LANGUAGES.includes(this.i18n.options.fallbackLng)
    ) {
      throw errObj({
        code: ErrCodes.CONFIG_IS_INVALID,
        params: {
          body: `'i18n.options.lng' must be included in [${SUPPORT_LANGUAGES.join(', ')}]`,
        },
      });
    }
  }
}

export const Rue = new RueRuntimeConfig$Base();
