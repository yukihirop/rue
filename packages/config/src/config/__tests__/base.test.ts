import { Config$Base as Config } from '../base';

describe('Config$Base', () => {
  describe('static properties', () => {
    it('should correctly', () => {
      expect(Config.fileName).toEqual('rue.config.js');
      expect(Config.default).toEqual({
        cli: {
          commands: {
            console: {
              nodeREPL: {
                prompt: '🍛 > ',
                useColors: true,
              },
            },
          },
        },
        repl: {
          loadModules: [
            'src/**/{forms,models,records}/**/*.{js,ts}',
            '!src/**/__tests__/*.test.{js,ts}',
          ],
          moduleAliases: { '@': './src' },
        },
      });
      expect(Config.defaultJS).toMatchSnapshot();
    });
  });

  describe('all', () => {
    describe('when default', () => {
      it('should correctly', () => {
        expect(Config.all()).toEqual({
          cli: {
            commands: {
              console: {
                nodeREPL: {
                  prompt: '🍛 > ',
                  useColors: true,
                },
              },
            },
          },
          repl: {
            loadModules: [
              'src/**/{forms,models,records}/**/*.{js,ts}',
              '!src/**/__tests__/*.test.{js,ts}',
            ],
            moduleAliases: { '@': './src' },
          },
        });
      });
    });
  });

  describe('nodeREPL', () => {
    describe('when default', () => {
      it('should correctly', () => {
        expect(Config.nodeREPL()).toEqual({
          prompt: '🍛 > ',
          useColors: true,
        });
      });
    });
  });

  describe('rueREPL', () => {
    describe('when default', () => {
      it('should correctly', () => {
        expect(Config.rueREPL()).toEqual({
          loadModules: [
            'src/**/{forms,models,records}/**/*.{js,ts}',
            '!src/**/__tests__/*.test.{js,ts}',
          ],
          moduleAliases: { '@': './src' },
        });
      });
    });
  });
});
