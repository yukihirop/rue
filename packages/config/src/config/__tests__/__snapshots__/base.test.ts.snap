// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config$Base static properties should correctly 1`] = `
"module.exports = {
  cli: {
    commands: {
      console: {
        nodeREPL: {
          prompt: '🍛 > ',
          useColors: true
        }
      }
    }
  },
  repl: {
    loadModules: [
      'src/**/{forms,models,records}/**/*.{js,ts}',
      '!src/**/__tests__/*.test.{js,ts}'
    ],
    moduleAliases: {
      '@': './src'
    }
  }
};"
`;
